File: examples/kotlin/cases/tasks/task_list/task_list.ui.case.kt
@file:Suppress("UnsafeCastFromDynamic")

package app.protocol.examples.kotlin.cases.tasks.task_list

import app.protocol.examples.kotlin.core.shared.AppHttpClient
import app.protocol.examples.kotlin.core.ui.BaseUiCase
import app.protocol.examples.kotlin.core.ui.UiContext
import kotlinx.browser.document
import kotlinx.coroutines.MainScope
import kotlinx.coroutines.launch
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.boolean
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import org.w3c.dom.HTMLElement

private val json = Json { ignoreUnknownKeys = false }
private val uiScope = MainScope()

data class ViewState(
    val tasks: List<Task> = emptyList(),
    val loading: Boolean = true,
    val error: String? = null,
)

data class ColumnViewModel(
    val status: String,
    val title: String,
    val tasks: List<Task>,
    val emptyMessage: String,
)

data class ViewModel(
    val columns: List<ColumnViewModel>,
    val feedback: String? = null,
)

class TaskListUi(
    ctx: UiContext,
) : BaseUiCase<ViewState>(
        ctx = ctx,
        initialState = ViewState(),
    ) {
    private val domainCase = TaskListDomain()

    override fun view(): Any {
        val root = document.createElement("section") as HTMLElement
        val designSystem = resolveDesignSystem()
        val renderCardActions = ctx.extra?.get("renderCardActions") as? ((Task) -> HTMLElement?)

        fun render() {
            root.innerHTML = ""
            val viewModel = _viewmodel(state) as ViewModel

            if (!viewModel.feedback.isNullOrBlank()) {
                val feedback = document.createElement("div") as HTMLElement
                feedback.className = "feedback feedback--error"
                feedback.textContent = viewModel.feedback
                root.appendChild(feedback)
            }

            val columns =
                viewModel.columns.map { column ->
                    val children =
                        if (column.tasks.isNotEmpty()) {
                            column.tasks.map { task ->
                                designSystem.taskCard(
                                    task.title,
                                    task.description,
                                    task.status,
                                    renderCardActions?.invoke(task),
                                ) as HTMLElement
                            }
                        } else {
                            listOf(
                                designSystem.emptyColumnState(column.emptyMessage) as HTMLElement,
                            )
                        }

                    designSystem.taskColumn(column.title, column.tasks.size, children) as HTMLElement
                }

            root.appendChild(
                designSystem.taskBoard(columns) as HTMLElement,
            )
        }

        render()
        uiScope.launch {
            loadTasks(::render)
        }

        return root
    }

    override suspend fun test() {
        val view = view()
        check(view is HTMLElement) {
            "test: view must return a visual unit"
        }

        val result = _service(TaskListInput) as TaskListOutput
        check(result.tasks.size == 2) {
            "test: ui service must return the mocked task list"
        }

        val viewModel =
            _viewmodel(
                ViewState(
                    tasks = result.tasks,
                    loading = false,
                    error = null,
                ),
            ) as ViewModel
        val todo = viewModel.columns.firstOrNull { it.status == "todo" }
        val doing = viewModel.columns.firstOrNull { it.status == "doing" }
        check(todo?.tasks?.size == 1 && doing?.tasks?.size == 1) {
            "test: ui viewmodel must group tasks by status"
        }
    }

    override fun _viewmodel(vararg args: Any?): Any {
        val current = args.firstOrNull() as? ViewState ?: state
        return ViewModel(
            columns = listOf(
                ColumnViewModel(
                    status = "todo",
                    title = "To Do",
                    tasks = current.tasks.filter { it.status == "todo" },
                    emptyMessage = if (current.loading) "Loading cards..." else "No cards in to do.",
                ),
                ColumnViewModel(
                    status = "doing",
                    title = "Doing",
                    tasks = current.tasks.filter { it.status == "doing" },
                    emptyMessage = if (current.loading) "Loading cards..." else "No cards in doing.",
                ),
                ColumnViewModel(
                    status = "done",
                    title = "Done",
                    tasks = current.tasks.filter { it.status == "done" },
                    emptyMessage = if (current.loading) "Loading cards..." else "No cards in done.",
                ),
            ),
            feedback = current.error,
        )
    }

    override suspend fun _service(vararg args: Any?): Any {
        domainCase.validate(TaskListInput)
        return _repository(TaskListInput)
    }

    override suspend fun _repository(vararg args: Any?): Any {
        val payload =
            resolveApiClient().request(
                mapOf(
                    "method" to "GET",
                    "url" to "/tasks",
                ),
            ) as? String ?: error("task_list.ui received an invalid list response")

        val envelope = json.parseToJsonElement(payload).jsonObject
        if (envelope["success"]?.jsonPrimitive?.boolean != true) {
            val message =
                envelope["error"]?.jsonObject?.get("message")?.jsonPrimitive?.content
                    ?: "Request failed"
            error(message)
        }

        val data = envelope["data"] ?: error("task_list.ui missing success payload")
        return json.decodeFromJsonElement(TaskListOutput.serializer(), data)
    }

    private suspend fun loadTasks(render: () -> Unit) {
        setState(state.copy(loading = true, error = null))
        render()
        try {
            val result = _service(TaskListInput) as TaskListOutput
            setState(
                ViewState(
                    tasks = result.tasks,
                    loading = false,
                    error = null,
                ),
            )
        } catch (error: Throwable) {
            setState(
                ViewState(
                    tasks = emptyList(),
                    loading = false,
                    error = error.message ?: "Failed to load tasks",
                ),
            )
        }
        render()
    }

    private fun resolveApiClient(): AppHttpClient =
        ctx.api ?: error("task_list.ui requires ctx.api")

    private fun resolveDesignSystem(): dynamic {
        val packages = ctx.packages ?: error("task_list.ui requires ctx.packages")
        return packages["designSystem"] ?: error("task_list.ui requires packages.designSystem")
    }
}

File: examples/go/cases/tasks/task_list/task_list.ui.case.go
package task_list

import (
	"encoding/json"
	"fmt"
	"html/template"

	"app-protocol/examples/go/core"
	"app-protocol/examples/go/core/shared"
)

type ViewState struct {
	Tasks   []Task
	Loading bool
	Error   string
}

type UI struct {
	support    core.UiCaseSupport[ViewState]
	domainCase *Domain
}

type CardActionRenderer func(task Task) (template.HTML, error)

type DesignSystemPackage interface {
	TaskBoard(columns []template.HTML) template.HTML
	TaskColumn(title string, count int, children []template.HTML) template.HTML
	TaskCard(title string, description string, status string, actions template.HTML) template.HTML
	EmptyColumnState(message string) template.HTML
	Feedback(kind string, message string) template.HTML
}

func NewUI(ctx *core.UiContext) *UI {
	return &UI{
		support: core.NewUiCaseSupport(ctx, ViewState{
			Tasks:   []Task{},
			Loading: true,
		}),
		domainCase: NewDomain(),
	}
}

func (u *UI) View() (template.HTML, error) {
	designSystem, err := u.resolveDesignSystem()
	if err != nil {
		return "", err
	}

	output, err := u.Service(TaskListInput{})
	if err != nil {
		return designSystem.Feedback("error", err.Error()), nil
	}

	renderCardActions, _ := u.support.Ctx.Extra["renderCardActions"].(CardActionRenderer)
	columns := []template.HTML{
		u.renderColumn(designSystem, "todo", "To Do", output.Tasks, renderCardActions),
		u.renderColumn(designSystem, "doing", "Doing", output.Tasks, renderCardActions),
		u.renderColumn(designSystem, "done", "Done", output.Tasks, renderCardActions),
	}

	return designSystem.TaskBoard(columns), nil
}

func (u *UI) Test() error {
	view, err := u.View()
	if err != nil {
		return err
	}

	if string(view) == "" {
		return fmt.Errorf("test: view must return a visual unit")
	}

	result, err := u.Service(TaskListInput{})
	if err != nil {
		return err
	}

	if len(result.Tasks) != 2 {
		return fmt.Errorf("test: ui service must return the mocked task list")
	}

	return nil
}

func (u *UI) Service(input TaskListInput) (TaskListOutput, error) {
	if err := u.domainCase.Validate(input); err != nil {
		return TaskListOutput{}, err
	}

	return u.repository(input)
}

func (u *UI) repository(_ TaskListInput) (TaskListOutput, error) {
	client := u.resolveAPIClient()
	if client == nil {
		return TaskListOutput{}, fmt.Errorf("task_list.ui requires ctx.api")
	}

	response, err := client.Request(map[string]any{
		"method": "GET",
		"url":    "/tasks",
	})
	if err != nil {
		return TaskListOutput{}, err
	}

	output, err := decodeTaskListOutput(response)
	if err != nil {
		return TaskListOutput{}, err
	}

	if err := u.domainCase.ValidateOutput(output); err != nil {
		return TaskListOutput{}, err
	}

	return output, nil
}

func (u *UI) resolveAPIClient() shared.AppHttpClient {
	if u.support.Ctx == nil {
		return nil
	}

	return u.support.Ctx.API
}

func (u *UI) renderColumn(designSystem DesignSystemPackage, status string, title string, tasks []Task, renderCardActions CardActionRenderer) template.HTML {
	filtered := []Task{}
	for _, task := range tasks {
		if string(task.Status) == status {
			filtered = append(filtered, task)
		}
	}

	children := []template.HTML{}
	if len(filtered) == 0 {
		children = append(children, designSystem.EmptyColumnState(fmt.Sprintf("Nenhum card em %s.", status)))
	} else {
		for _, task := range filtered {
			var actions template.HTML
			if renderCardActions != nil {
				rendered, err := renderCardActions(task)
				if err == nil {
					actions = rendered
				}
			}

			children = append(children, designSystem.TaskCard(
				task.Title,
				task.Description,
				string(task.Status),
				actions,
			))
		}
	}

	return designSystem.TaskColumn(title, len(filtered), children)
}

func (u *UI) resolveDesignSystem() (DesignSystemPackage, error) {
	if u.support.Ctx == nil || u.support.Ctx.Packages == nil {
		return nil, fmt.Errorf("task_list.ui requires packages.designSystem")
	}

	designSystem, ok := u.support.Ctx.Packages["designSystem"].(DesignSystemPackage)
	if !ok || designSystem == nil {
		return nil, fmt.Errorf("task_list.ui requires packages.designSystem")
	}

	return designSystem, nil
}

func decodeTaskListOutput(value any) (TaskListOutput, error) {
	var output TaskListOutput
	encoded, err := json.Marshal(value)
	if err != nil {
		return output, err
	}

	if err := json.Unmarshal(encoded, &output); err != nil {
		return output, err
	}

	return output, nil
}
